### ISSUE TITLE
Computing ephemerides from orbits

### ISSUE THREAD

**Issue by s-m-e** [poliastro/poliastro#1008]
I am trying to compute ephemerides from orbits with poliastro. Bluntly put, I need Cartesian coordinates in fixed time intervals for longer periods of times. poliastro can generate `Orbit` objects from ephemerides objects, but - as far as I can see - there is no direct method for going the opposite direction (in some optimized form).

What I came up with simply looks as follows ... I am not entirely sure if this the "right" way to do this:

```python
eph = np.zeros((days, 3))
for day in range(0, days):
    time_current = time_zero + day * u.day
    eph[day, :] = orb.propagate(time_current).r.to(u.AU)
```

The approach is relatively slow.

**Comment by astrojuanlu** [comment]
Hi @s-m-e :smiley: 

> Bluntly put, I need Cartesian coordinates in fixed time intervals for longer periods of times. 

Two options, depending on what exactly you want to do:

- Use `Ephem` objects, like described in the [User Guide](https://docs.poliastro.space/en/latest/user_guide.html#where-are-the-planets-computing-ephemerides) ([reference](https://docs.poliastro.space/en/latest/autoapi/poliastro/ephem/index.html#poliastro.ephem.Ephem))
- Instead of a `for` loop to call `Orbit.propagate`, use `from poliastro.twobody.propagation import propagate` and pass the initial orbit and the array of times, like done in the [Cowell formulation notebook](https://docs.poliastro.space/en/latest/examples/Propagation%20using%20Cowells%20formulation.html) or the [Perturbations notebook](https://docs.poliastro.space/en/latest/examples/Natural%20and%20artificial%20perturbations.html)

Do any of these two options satisfy your needs?

**Comment by astrojuanlu** [comment]
> The approach is relatively slow ... probably due to lots of overhead from (numba) function calls (?).

Most likely, it's a combination of excessive unit conversion (see #1006) and excessive cartesian - classical roundtripping (see #921).

**Comment by s-m-e** [comment]
Thanks @astrojuanlu. I am working with "classical" orbital elements. I am looking into ways of scaling the process up to the currently known ~1M asteroids. JPL's Horizons computes ephemerides "on demand". I.e. using their system would result in an insane load at their end plus some serious internet traffic, so I can rule `Ephem.from_horizons` more or less out (although I'd love to have a "dump" of their data at their incredible level of precision ...). Currently, my primary objective is testing a model which does not need very accurate data - at least for the testing bit.

A couple of hours of experimenting later ... `from poliastro.twobody.propagation` is the key to considerably more performance. The solution looks about as follows:

```python
time_vector = Time([time_zero + day * u.day for day in range(0, days)])
timedelta_vector = time_vector - orb.epoch
eph = np.zeros((days, 3)) # actually a larger data structure ...
eph[:, :] = np.transpose(propagate(orb, timedelta_vector).xyz.to(u.AU).value)
```

For `days == 36525`, this results in a 128x speedup compared to my initial approach (!). So far, I have used poliastro's default propagator, `farnocchia`. I am wondering if I can gain some extra performance by trading accuracy and going for another propagator.

**Comment by astrojuanlu** [comment]
Glad it worked! Curious to know about the "two hours of experimentation" in case there's something we can improve or fix in our docs :) 

We have not benchmarked our propagators (noted! https://github.com/poliastro/benchmarks/issues/21) but I would say that `farnocchia` is the simplest.

Potentially, there's more performance gain that could be achieved because, if you already have classical orbital elements, you could avoid the `rv2coe` call here:

https://github.com/poliastro/poliastro/blob/c4d0f80c98fb2a0ec59a07f4664cfebb0b1a6cf9/src/poliastro/core/propagation/farnocchia.py#L319

The `coe2rv` call later is necessary to get the `xyz` position, which is what you're interested in. Therefore, for maximum performance, without thinking about it very much, I would do something like this:

```python
@jit
def farnocchia_coe(k, p, ecc, inc, raan, argp, nu0, tof):
    q = p / (1 + ecc)

    delta_t0 = delta_t_from_nu(nu0, ecc, k, q)
    delta_t = delta_t0 + tof

    nu = nu_from_delta_t(delta_t, ecc, k, q)

    return coe2rv(k, p, ecc, inc, raan, argp, nu)

# orb classical orbital elements already known
positions = [
    farnocchia_coe(k, p, ecc, inc, raan, argp, nu0, tof)[0]
    for tof in timedelta_vector
]  # km, you can convert to `u.AU` later if needed
```

I anticipate that this would give you a decent boost (not 128x, but noticeable anyway).

**Comment by s-m-e** [comment]
> Curious to know about the "two hours of experimentation" in case there's something we can improve or fix in our docs :)

Your docs are good. I was simply "missing" some design concepts, i.e. how certain classes / objects work and interact. It's sort of intuitive, but what you guys implemented was still beyond my imagination. I ended up reading the source code.

Actually there is one thing "missing" from my perspective, but not in the docs: A clean, fast and out-of-the-box way of computing an `Ephem` object from an `Orbit` object, e.g. `Ephem.from_orbit(orb, time_stamps, propagator, ...)`.

> I would say that farnocchia is the simplest.

It is the fastest, more or less, yes. But all except one, if I remember correctly, are within another 10% to 20% of farnocchia's performance, so - actually to my surprise - there are no significant differences.

> 128x

This number comes from an aged 3rd-gen i7. On a 2nd-gen AMD Epyc with AVX2 support, the speedup is even greater. 

> Potentially, there's more performance gain that could be achieved because, if you already have classical orbital elements, you could avoid the rv2coe call here [...]

Interesting. `rv2coe` looks rather expensive, yes. I'll test this and report back.

**Comment by s-m-e** [comment]
Ok, the implementation looks straight forward, but `numba` (0.51.1) keeps throwing errors at me. Maybe you understand what is going on:

```python
from poliastro.core._jit import jit
from poliastro.core.elements import coe2rv
from poliastro.core.propagation.farnocchia import delta_t_from_nu, nu_from_delta_t

from astropy.coordinates import CartesianDifferential, CartesianRepresentation

@jit
def farnocchia_coe(k, p, ecc, inc, raan, argp, nu0, tof):
    
    q = p / (1 + ecc)

    delta_t0 = delta_t_from_nu(nu0, ecc, k, q)
    delta_t = delta_t0 + tof

    nu = nu_from_delta_t(delta_t, ecc, k, q)

    return coe2rv(k, p, ecc, inc, raan, argp, nu)

def propagate_fast(orbit, time_of_flight, *, method=farnocchia_coe, rtol=1e-10, **kwargs):

    rr, vv = farnocchia_coe(
        orbit.k,
        orbit.p,
        orbit.ecc,
        orbit.inc,
        orbit.raan,
        orbit.argp,
        orbit.nu,
        time_of_flight.reshape(-1).to(u.s),
    )
    
    # TODO: Turn these into unit tests
    assert rr.ndim == 2
    assert vv.ndim == 2

    cartesian = CartesianRepresentation(
        rr, differentials = CartesianDifferential(vv, xyz_axis = 1), xyz_axis = 1
    )

    return cartesian
```

**Comment by astrojuanlu** [comment]
Yes, I think you skipped some steps: `farnocchia_coe` being a numba-accelerated function cannot be called with astropy `Quantity` objects, nor with an array of times of flight. There's an intermediate function missing:

```python
def farnocchia_coe_safe(k, p, ecc, inc, raan, argp, nu0, tofs):
    k = k.to(u.km ** 3 / u.s ** 2).value
    p = p.to_value(u.km)
    ecc = ecc.value
    inc = inc.to_value(u.rad)
    raan = raan.to_value(u.rad)
    argp = argp.to_value(u.rad)
    nu0 = nu0.to_value(u.rad)
    tofs = tofs.to(u.s).value

    results = [farnocchia_coe(k, p, ecc, inc, raan, argp, nu0, tof) for tof in tofs]
    # TODO: Rewrite to avoid iterating twice
    return (
        [result[0] for result in results] * u.km,
        [result[1] for result in results] * u.km / u.s,
    )
```

And also, it should be `orbit.attractor.k` in your `propagate_fast`.

With those two changes, it works!

```
In [22]: from poliastro.examples import iss                                                                                                                                                                                                   

In [23]: from poliastro.util import time_range                                                                                                                                                                                                

In [24]: tofs = time_range(start=iss.epoch, end=iss.epoch + 3 * u.h, periods=10) - iss.epoch                                                                                                                                                  

In [25]: propagate_fast(iss, tofs)                                                                                                                                                                                                            
Out[25]: 
<CartesianRepresentation (x, y, z) in km
    [(  859.07256   , -4137.20368   ,  5295.56871   ),
     ( 6553.04951593,   924.32778774,  1500.28797426),
     ( 1941.43870265,  4538.04577024, -4661.37600883),
     (-5716.57707911,  1022.02289903, -3498.02405503),
     (-4359.67616826, -4097.72461402,  3172.0763872 ),
     ( 3884.55923795, -2743.48905788,  4830.42173798),
     ( 6013.21594065,  2941.43139779, -1129.71408649),
     (-1308.43010489,  4007.11773564, -5318.57856068),
     (-6564.62868167, -1225.44261669, -1144.34637306),
     (-1460.83036048, -4518.52703816,  4829.20937759)]
 (has differentials w.r.t.: 's')>
```

**Comment by s-m-e** [comment]
I totally missed the wrapper function around `farnocchia` in `src/poliastro/twobody/propagation.py`, yes. I also expected that quantities were transparently handled as numpy arrays / scalars. Thanks for the explanation.

Ok, this approach works, but it runs two times longer than the [original "fast" approach](https://github.com/poliastro/poliastro/issues/1008#issuecomment-685662136), so it's actually slower.

**Comment by astrojuanlu** [comment]
Oh wow, that's... surprising. If you share a self-contained script where both approaches are implemented, I can give a hand with some profiling.

**Comment by s-m-e** [comment]
Ok, this was a numba issue if I understand this correctly. I uploaded a [self contained notebook here](https://gist.github.com/s-m-e/c7a5e5a09df54ffdb77003d423cf603b). I "fixed" it by separating the contents of cells 6 and 12. If this stuff is on one single cell, it probably re-compiles my `farnocchia_coe_fast` function probably on every run.

It's actually faster, from 530 ms down to 390 ms, so roughly 25% less computing time.

**Comment by astrojuanlu** [comment]
Amazing! I think any other improvement will be smaller than that, although I could be surprised. Do you plan to keep optimizing this code?

**Comment by s-m-e** [comment]
I am wondering if I can gain anything by scaling "horizontally": Propagating multiple orbits simultaneously, without iteration. I have another problem where I do not need lots of coordinates per orbit, just one, but from 1M orbits in one go. Any ideas?

**Comment by astrojuanlu** [comment]
I think plain `multiprocessing` should work well in this case, all the orbits are independent to one another and the propagation itself is serial so it runs on a single core. Haven't tried it personally, but the relevant objects are pickleable and it should be feasible.

**Comment by s-m-e** [comment]
I am already running my code with `multiprocessing` on high-core-count machines, but this is more like a brute-force way of improving the performance - and it still could be much better. Calling `orb.propagate(some_time).r.to(u.AU)` once per 1M orbits is just as slow as calling `orb.propagate(time_step).r.to(u.AU)` for one million points on a single orbit (via `for` loops). I'd also expect some significant gains by exploiting vectorization, i.e. getting rid of the `for` loops in this case as well. The question is if `poliastro.twobody.propagation.propagate` can be rewritten in such a way that it accepts an "array of orbit objects", vectorizing instructions across all passed orbits wherever possible.

**Comment by astrojuanlu** [comment]
You're right, the `for` loop happens in the list comprehension of `farnocchia_coe_safe` (as happens with all the propagators), so by rewriting that part in numba we could potentially save some milliseconds. This is the subject of #922.

On the other hand:

> The question is if poliastro.twobody.propagation.propagate can be rewritten in such a way that it accepts an "array of orbit objects", vectorizing instructions across all passed orbits wherever possible.

Yes, that's a very good question indeed. Perhaps we could introduce a new function, `propagate_many` or something like that, taking care of it. I assume that receiving a list of `Orbit` objects could be the first step, but maybe we can think of better ways of achieving the same thing.

**Comment by s-m-e** [comment]
> Yes, that's a very good question indeed. Perhaps we could introduce a new function, propagate_many or something like that, taking care of it. I assume that receiving a list of Orbit objects could be the first step, but maybe we can think of better ways of achieving the same thing.

I have this idea in the back of my head ... and your code looks like it could work: Would you consider a pull request based on [numba's integration of CUDA](https://numba.pydata.org/numba-doc/dev/cuda/index.html)?

**Comment by astrojuanlu** [comment]
Aw yeah, absolutely! :grinning:

**Comment by s-m-e** [comment]
I experimented with `numba`'s implementation of CUDA ... Well, writing bare C CUDA kernels is almost equally frustrating.

- `numba.cuda.jit` does not support memory allocations in functions, see [here](https://numba.pydata.org/numba-doc/dev/cuda/cudapysupported.html#numpy-support), which **drastically** limits `numpy` support. `poliastro` allocates memory through `numpy.array` virtually all over the place. This requires "a bit" of a re-design. 

> Due to the CUDA programming model, dynamic memory allocation inside a kernel is inefficient and is often not needed. Numba disallows any memory allocating features. This disables a large number of NumPy APIs.

- `numba.cuda.jit` does not like slices or any form of Python's vector notation. Back to for loops, everywhere.

- `numba.cuda.jit` does not support `matmul`, see [here](https://github.com/numba/numba/issues/3804) and [here](https://github.com/numba/numba/issues/3409). For the multiplication of large matrices, [exploiting CUDA by implementing the multiplication manually makes sense](https://numba.pydata.org/numba-doc/dev/cuda/examples.html). However, functions like `coe_rotation_matrix` in `poliastro` rely on small 3x3 matrices. A dumb for-loop based implementation is required.

- Default values for parameters are completely ignored. In function calls, keyword arguments are causing issues - keywords have to be removed.

It's theoretically possible to write functions that are compatible with both`numba.njit` and `numba.cuda.jit` but they may look incredibly ugly. Having a lot of functions twice on the other hand, for CPUs and GPUs, also makes stuff a lot less maintainable. Not sure where to go.

My first test implementation can be found [here](https://github.com/pleiszenburg/bulk_lambert/blob/898f9d84dea57f17ef75fb2a138f29e14198aeb5/bulk_lambert/cuda/bulk.py#L24). It *should* produce the same result as [this CPU-based and validated function](https://github.com/pleiszenburg/bulk_lambert/blob/898f9d84dea57f17ef75fb2a138f29e14198aeb5/bulk_lambert/bulk.py#L27) - however, it does not yet. I guess some index, sign or trigonometry issue ...

The actually interesting part: My very aged Quadro K1000M outperforms a single core of my also aged i7 3740QM by about 6%. It's ... a good sign ... but there is room for improvement ;)

**Comment by s-m-e** [comment]
> Yes, that's a very good question indeed. Perhaps we could introduce a new function, propagate_many or something like that, taking care of it. I assume that receiving a list of Orbit objects could be the first step, but maybe we can think of better ways of achieving the same thing.

I also looked into this, i.e. "horizontal scaling". A "naive" [test implementation can be found here](https://github.com/pleiszenburg/bulk_lambert/blob/3a1df9f19ab47076400b176c0c685ebb40fe3f47/bulk_lambert/bulk_many.py#L26). It's based on a "list of Orbit objects" as you suggested. Playing with 20k orbits at a time, this approach causes an increase in performance of about 10 times (relative to poliastro's current implementation). It's some mild vectorization at the highest level plus mildly optimized removal of units plus a little extra JIT. I guess I could squeeze out 20 or 30 times, but this approach is a dead end as far as I am concerned. If the orbits (and their internal states) were already organized as arrays, you can easily gain 400 times (relative to poliastro's current implementation). I suggest to add array types matching the following classes and relevant parts of their APIs: `Orbit`, `BaseState`, `ClassicalState` and `ModifiedEquinoctialState`.

I have also looked into vectorizing the entire stack below the propagators. The problem is that there are tons of edge cases. It's not trivial at all to make this work (and faster). For very "similar" orbits, it's easy - maybe "grouping" individual orbits by categories within an array makes sense.

**Comment by astrojuanlu** [comment]
Catching up with notifications, and writing publicly some things @s-m-e and I discussed in private.

So, first things first: it's **awesome** that we can leverage GPUs for this :rocket: Let's slowly move towards a roadmap where this is both possible and maintainable.

A few ideas:

> Having a lot of functions twice on the other hand, for CPUs and GPUs, also makes stuff a lot less maintainable. Not sure where to go.

As part of #921 we also wanted to get deeper into numba and vectorize the loops themselves. After having a glimpse at the GPU code that @s-m-e showed me, it was clear that we will need some amount of duplication to support all the use cases without making the code super complicated. However, I wonder if some parts can be extracted, like "kernels", that can be reused for all of them.

> poliastro allocates memory through numpy.array virtually all over the place. This requires "a bit" of a re-design.

Indeed. When numba added support for this we got relaxed, but perhaps we should backtrack, and rewrite the `poliastro.core` functions in a way that they don't allocate memory. I wonder if there is some keyword to `numba.njit` that makes the compilation fail if any memory allocation is taking place. If not, we'll have to do it manually.

> numba.cuda.jit does not support matmul, see here and here.

Reading https://github.com/numba/numba/issues/3409#issuecomment-429814363, I wonder if we should think about using CuPy for this. But I can't even imagine the implications of that, never used CuPy myself. If not, manual matrix multiplication as you say might be a good enough backup plan.

> I suggest to add array types matching the following classes and relevant parts of their APIs: Orbit, BaseState, ClassicalState, ModifiedEquinoctialState and RVState. States (and therefore Orbits) within an array should have a common attractor and a common plane, I think.

This is an appealing proposal. I had never considered this in the poliastro roadmap, but as discussed, it enables lots of performance optimizations. I think it's similar to what happened with Shapely and the new PyGEOS ([source](https://jorisvandenbossche.github.io/talks/2020_GeoPython_Shapely-2.0)).

> I have also looked into vectorizing the entire stack below the propagators (for extra performance beyond the mentioned 400 times). The problem is that there are tons of edge cases. It's not trivial at all to make this work (and faster). For very "similar" orbits, it's easy - maybe "grouping" individual orbits by categories within an array makes sense.

I had this idea for a project with Libre Space Foundation but I haven't had time to get to it yet. I will start working on this very soon though. I wanted to apply it only for SGP4 but maybe it's worth it for other propagators as well.

---

Going back to the idea of the "kernels": I think poliastro nowadays serves as a nice "reference implementation" of several propagation algorithms, low thrust guidance laws, and such. The moment we start trying to break these down into kernels and introducing micro optimizations, this readability will be lost. I would like to have a plan on how do we plan to tackle this: some sort of "poliastro-lite" library? "The Museum of Astrodynamics Algorithms"? "The Rosetta Stone of Propagators"? It would be awesome to preserve them somehow before we start making poliastro ultra fast at the cost of interpretability and maintainability. Also because these reference versions are much closer to the original paper or book, simpler to modify, and could be used to generate "ground truth" for tests.

---

This discussion is enough for one year of full-time work, maybe more. Which means that we need a plan so we don't get too impatient, because we don't have full time "employees" just yet :)

**Comment by SidharajYadav** [comment]
How i contribute this issues

**Comment by astrojuanlu** [comment]
Hi @SidharajYadav, thanks for your interest! However, @s-m-e is already working on this.

**Comment by Ahsoka** [comment]
Hi, any updates on this?

**Comment by astrojuanlu** [comment]
@Ahsoka Unfortunately no, any updates will be posted on this issue. How are you affected by this? What's your use case?

**Comment by s-m-e** [comment]
> If I may inquire what happened with the NumFOCUS grant? Did the funding get cut?

Nope. Most of the work has been carried out, i.e. analysis, design, prototypes, etc. - with substantial speedups possible as expected. Though, it requires changes touching many areas of the package. This ultimately never got integrated into the package itself, however, which is a longer story. @astrojuanlu steeped down as maintainer a while ago. Now it's technically up to he new maintainer @jorgepiloto to decide how to proceed, though I have not heard from him in a while, too. As for my part, I have been sitting this one out lately, considering to fork before this work gets lost forever provided I find the extra time.

Gentlemen, there is apparently demand - what's your current position in this matter?

**Comment by astrojuanlu** [comment]
@jorgepiloto might not be able to interact with this repository in a very long time for personal reasons. The future of poliastro doesn't look very bright and forking might be the best course of action for now.

**Comment by Ahsoka** [comment]
@s-m-e Where can I find this code?  In case I wanted to experiment with it. :)

**Comment by s-m-e** [comment]
@Ahsoka As a point of entry based on what you appear to be looking for, I can recommend [this notebook](https://github.com/pleiszenburg/bulk_propagate/blob/master/test_guv_farnocchia.ipynb). 

Most experiments reside in [this repo](https://github.com/pleiszenburg/bulk_lambert) and in [this repo](https://github.com/pleiszenburg/bulk_propagate). An incomplete integration into `poliastro.core` can be found in a [branch of my fork of poliastro](https://github.com/s-m-e/poliastro/tree/feature_jit-core). The above mentioned notebook is built around a certain older commit in this fork which can be used for testing and benchmarking (indicated at the top of the notebook).
